[1단계 - 페이먼츠 모듈 ] 제나(이정은) 미션 제출 합니다.

JeLee-river opened the pull request:
안녕하세요 유조! 자동차 경주 미션에 이어 다시 만나 뵙게 되어 정말 반갑습니다!! 😁 이번 미션도 잘 부탁드립니다. ㅎㅎ

---

## 🎯 페이먼츠 모듈

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계 - Module

- 재사용 가능한 모듈화된 컴포넌트를 개발하고 npm에 배포할 수 있다.
- 재사용 가능한 커스텀 훅을 개발하고 npm에 배포할 수 있다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 문서화하고 테스트할 수 있다.
- React Testing Library를 사용하여 컴포넌트와 커스텀 훅에 대해 폭넓은 테스트를 작성할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항(모달 컴포넌트 및 커스텀 훅 개발, npm 배포)을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] Storybook과 RTL 테스트 케이스를 모두 작성했나요?
- [x] 코드 컨벤션을 지키고, 에러 핸들링을 고려했나요?
- [ ] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 기본 배포 링크 기입: **\_\_**
- [x] 배포한 npm 패키지가 정상적으로 설치 및 작동하나요?
  - 배포 npm 링크: 
    - **[컴포넌트](https://www.npmjs.com/package/@leeje/modal-component-module)**
    - **[커스텀 훅](https://www.npmjs.com/package/@leeje/hooks-module)**
- [x] Storybook 데모 페이지에 정상적으로 접근할 수 있나요?
  - Storybook 링크: **[링크](https://6800ba1c9f8c4a9060a108bb-itqiqsllzf.chromatic.com/)**

### 리뷰 요청 & 논의하고 싶은 내용

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

1. 이번 단계에서는 사용자의 자유도를 어디까지 허용해야 할지 가장 많이 고민했습니다.
예를 들어, 모달 컴포넌트에서는 기능 요구사항 외에 모달 내부의 UI요소 배치, UI 스타일, 버튼 추가 여부를 사용자가 정하도록 할지 고민하였고, 기본 옵션을 제공하되, 사용자가 원하는 경우 커스텀할 수 있도록 구현했습니다. 

이 과정에서 '모듈로서의 일관성을 유지하면서도 적절한 자유도를 보장하는 기준'을 세우는 일이 쉽지 않았습니다. 이에 따라 페어와 계속해서 '어느 정도까지 사용자가 설정을 조정해도 괜찮을까?'를 주제로 논의하며 방향을 잡아 나갔습니다.

<br>

2. 커스텀 훅을 설계하면서 미션 요구사항에 따른 구현 방향을 정하는데 고민이 있었습니다. 
저희 페어는 '컴포넌트에서 React 훅을 사용한 로직을 분리한 코드' 를 커스텀 훅이라 정의했습니다. 그러나 이번 미션의 핵심은 '카드 정보의 유효성 검사 결과와 에러 메시지를 사용자에게 제공하는 것'이었고, 이 로직은 훅이 아닌 단순 유틸 함수로도 구현할 수 있지 않을까 하는 의문이 들었습니다.

따라서 최종적으로는 커스텀 훅이 갖는 상태 관리의 장점을 살리고자, 카드 번호나 유효 기간 등의 입력 값을 훅 내부에서 상태로 관리하고, 검증 결과(isValid)와 에러 메시지(errorMessage)를 함께 반환하는 구조로 결정했습니다.

<br>
<br>

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

### 2 - 1. 배포 목적인 모듈의 자유도

배포 목적으로 구현하는 모듈의 자유도를 정하는 의사 결정 과정이 궁금합니다.

이번에 처음으로 모듈을 개발하며 사용자의 자유도를 어디까지 제한해야 할지 정하는데 어려움이 있었습니다. 사용자의 자유도를 지나치게 허용하면 모듈로서의 의미가 사라질 것 같고, 자유도를 제한하면 활용성이 떨어진다고 생각하여 적절한 기준에 대해 고민했습니다. 제가 생각했을때 이러한 사용자의 자유도는 배포 초기에는 확정 짓기 어려우며 지속적인 업데이트로 바꿔갈 수 있다고 결론을 내렸고, 이에 따라 초기 버전에서는 일관된 기능으로 배포해도 괜찮지 않을까 생각했는데 이에 대한 유조의 의견이 궁금합니다.

<br>

### 2 - 2. 리팩토링

사용성, 기능 확장성 측면에서 리팩토링하면 좋을 부분이 있을지 궁금합니다.
지금까지 겪었던 미션과 다르게 이번 미션은 사용자가 개발자인 모듈인 점에서, 사용성과 기능 확장에 대한 유연한 대응이 중요할 것 같다고 느꼈습니다. 혹시 이러한 측면에서 리팩토링하면 좋을 부분이나, 고려해야할 요소가 있을지 유조의 의견을 여쭙고 싶습니다.

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 모달 컴포넌트의 유연성

- 다양한 위치(중앙, 하단 등)와 내용을 prop으로 받아 모달을 유연하게 구성할 수 있는가?
- 열기, 닫기, 확인 등의 사용자 정의 이벤트 핸들러가 자연스럽게 동작하는가?
- Storybook을 통해 다양한 시나리오를 명확히 테스트했는가?

### 2. 커스텀 훅의 역할과 모듈화

- 카드 정보 입력 유효성 검사 훅이 각각 명확한 책임을 가지도록 분리되어 있는가? 
- 훅이 유효성 검사 결과와 에러 메시지를 명확히 반환하는가?
- 독립적인 테스트 케이스를 작성하여 정상/비정상 입력 모두를 폭넓게 검증했는가?

### 3. 모듈 배포
- npm 배포를 위한 설정이 적절히 구성되어 있는가?

jaeyoung-kwon committed:
docs: 기능 목록 작성

jaeyoung-kwon committed:
chore: prettier 설정

jaeyoung-kwon committed:
feat: 모달 열고 닫기 기능 구현

JeLee-river committed:
feat: 모달을 열었을 때 dim 처리되는 기능 구현

jaeyoung-kwon committed:
feat: 모달을 열었을 때 배경을 클릭하면 모달이 닫히는 기능 구현

JeLee-river committed:
feat: 모달 내부 콘텐츠를 사용자에게 받는 기능 구현

jaeyoung-kwon committed:
feat: 모달 위치를 사용자한테 받는 기능 구현

JeLee-river committed:
feat: 모달 제목을 사용자에게 받는 기능 구현

JeLee-river committed:
feat: 닫기 버튼 표시 여부를 사용자에게 받는 기능 구현

jaeyoung-kwon committed:
feat: reset css 적용 및 메인 버튼에 대한 기능을 사용자에게 받는 기능 구현

JeLee-river committed:
feat: 보조 버튼을 사용자에게 받는 기능 구현

JeLee-river committed:
style: 모달의 position이 bottom일때 스타일 추가

jaeyoung-kwon committed:
refactor: 모달 버튼 컴포넌트화 및 모달 props로 받을 수 있게 변경

JeLee-river committed:
test: 모달 컴포넌트의 storybook 작성

jaeyoung-kwon committed:
style: emotion 라이브러리 적용

jaeyoung-kwon committed:
build: npm 배포 설정

JeLee-river committed:
feat: 카드번호의 유효성을 검증하는 useCardNumbers 커스텀 훅 생성

jaeyoung-kwon committed:
chore: prettier 설정

jaeyoung-kwon committed:
refactor: cardNumber 디렉터리 구조화

jaeyoung-kwon committed:
feat: 유효기간의 유효성을 검증하는 useExpiryDate 커스텀 훅 생성

JeLee-river committed:
feat: cvc 번호의 유효성을 검증하는 useCVC 커스텀 훅 생성

JeLee-river committed:
refactor: 커스텀 훅의 확장자명을 ts로 변경

JeLee-river committed:
test: useCardNumber 테스트 코드 추가

JeLee-river committed:
test: useCVC 테스트 코드 추가

JeLee-river committed:
test: useExpiryDate 테스트 코드 추가

JeLee-river committed:
fix: 만료된 유효기간을 검증하는 함수의 로직이 잘못 동작하는 문제 개선

jaeyoung-kwon committed:
refactor: 모달 컴포넌트 컴파운드 컴포넌트 패턴으로 리팩토링

JeLee-river committed:
test: Modal 컴포넌트의 변경에 따른 storybook 코드 수정

JeLee-river committed:
refactor: isOpen을 open으로 변수명 변경

jaeyoung-kwon committed:
feat: 사용자 지정 스타일 적용할 수 있도록 기능 추가

jaeyoung-kwon committed:
refactor: 하나의 errorMessage 상태를 케이스별로 상태 분리

JeLee-river committed:
feat: 사용자가 카드 번호 입력값을 제한할지 정하는 기능 추가

jaeyoung-kwon committed:
refactor: useCardNumbers 커스텀훅 변경을 테스트 코드에 반영 및 상수 분리

jaeyoung-kwon committed:
feat: 사용자가 유효기간 및 CVC 입력값을 제한할지 정하는 기능 추가

jaeyoung-kwon committed:
build: npm 배포 설정

jaeyoung-kwon committed:
build: components npm 배포 정보 수정

JeLee-river committed:
docs: npm 리드미 작성

JeLee-river committed:
docs: modal 컴포넌트의 npm 리드미 작성

yujo11 reviewed (changes requested):
안녕하세요 제나 ~~ 다시 만나서 반갑습니다 이번 미션도 잘 부탁드려요 ㅎㅎ

전체적을 일관성 있고 짜임새 있는 코드를 작성해주셔서 코멘트를 남길 부분이 많지 않았는데요. 코드를 보면서 조금 더 개선 시켜볼만한 부분과 궁금했던 부분들에 코멘트를 남겼으니 확인해주시면 좋을거 같습니다!

>배포 목적인 모듈의 자유도

제나가 말씀해주신 것처럼  일관성과 확장성 사이의 균형을 잘 잡아야 하는 참 어려운 부분인데요. 이 부분은 배포 목적의 모듈 뿐만 아니라 하나의 코드베이스에서 공통으로 사용되는 모듈들을 구현 할 때도 같은 고민을 하게 될거에요 

초기 버전에 확실한 사용 사례에 집중해서 일관된 기능으로 배포하는 것도 좋은 접근 방법이라고 생각합니다. 유저의 요구사항에 맞게 점진적으로 자유도를 확장 시켜나가면 좋을거 같아요 . 지금 구현해주신 모듈들의 형태도 이미 좋은 균형점에 위치해있다고 생각합니다 ㅎㅎ

많이 사용되는 디자인 시스템 라이브러리들에서는 어떤 부분을 기본적으로 제공하고, 어떤 부분을 커스텀 할 수 있도록 열어두는지도 참고해보시면 좋을거 같네요!

- https://ant.design/components/modal
- https://mui.com/material-ui/react-modal/

> 리팩토링

`useCardNumbers`, `useCVC` 등에서 유사한 코드 패턴이 반복되고 있는데요. 이 부분을 추상화해서 사용해 보는걸 생각해 볼 수 있을거 같아요. "사용자가 개발자인 모듈"이라는 사실은 개인적으로는 코드베이스 레벨에서 크게 중요하지 않다고 생각하는데요. 일반적인 애플리케이션의 경우도 한번 개발하고 끝나는게 아니라 유저의 요구사항, 개발자의 의도에 따라 계속해서 기능이 추가되고, 이런 기능 추가를 쉽게 할 수 있는 코드베이스를 만들도록 신경 쓰는데요. 같은 관점에서 유지보수하기 좋은 코드베이스를 만들어나가신다면 새로운 요구사항들에 보다 쉽게 대응하실 수 있을거라고 생각합니다 ㅎㅎ

----

미션 진행하느라 고생 많으셨습니다 제나 ~~ 👍 👍 👍

imakerjun reviewed (commented):
@coderabbitai Based on commit history, identify the **recurring work patterns** this developer uses to solve problems.

Focus on these two key questions:

1. **What consistent workflow or cycle does the developer follow?**
    
    – For example: feature → styling → testing. Explain with specific commit examples.
    
2. **What does this pattern suggest about the developer’s mindset or strategy?**
    
    – e.g., feature-first, cleanup-later, test-driven, etc.
    

Structure the result with a **clear title + example + interpretation**, and provide the answer **in Korean**.

wmakerjun commented:
@coderabbitai Based on commit history, identify the **recurring work patterns** this developer uses to solve problems.

Focus on these two key questions:

1. **What consistent workflow or cycle does the developer follow?**
    
    – For example: feature → styling → testing. Explain with specific commit examples.
    
2. **What does this pattern suggest about the developer’s mindset or strategy?**
    
    – e.g., feature-first, cleanup-later, test-driven, etc.
    

Structure the result with a **clear title + example + interpretation**, and provide the answer **in Korean**.

JeLee-river committed:
chore: prettier 설정 파일을 최상위 디렉터리로 이동

- 하나의 파일로 prettier 설정을 할 수 있도록 구조 변경

JeLee-river commented on components/package.json:
```diff
@@ -1,7 +1,7 @@
-  "version": "0.0.1",
-  "description": "npm modules",
+  "name": "@jae-o/modal-component-module",
+  "version": "0.0.4",
```
버전의 경우, 저희가 배포 중에 여러 번 시행착오를 겪으면서 의도치 않게 올라가게 되었습니다. 😂

JeLee-river commented on components/.prettierrc:
prettier config 파일의 위치를 최상위 폴더(react-modules)로 옮겨, 하나의 설정 파일로 줄였습니다! package.json 없이도 prettier config가 인식된다는 점을 처음 알게 됐습니다. 감사합니다 ㅎㅎ

JeLee-river replied:
prettier 설정 파일을 components, hooks의 상위 디렉터리로 이동시켜 하나의 설정 파일만 사용하도록 개선했습니다!

[chore: prettier 설정 파일을 최상위 디렉터리로 이동 (3498ec5)](https://github.com/woowacourse/react-modules/pull/114/commits/ce3c16072b6d6d54be89e762d7f8e9d105ecd4b4)

JeLee-river commented on components/package.json:
```diff
@@ -18,6 +18,8 @@
     "build-storybook": "storybook build"
   },
   "dependencies": {
+    "@emotion/react": "^11.14.0",
+    "@emotion/styled": "^11.14.0",
```
사실 처음에는 CSS 모듈을 사용했습니다. 페어와 저 모두 이를 다루어 본 경험이 있고 css를 컴포넌트 파일에서 분리해 관리하면 코드가 길어지지 않고 가독성이 좋은 것이 장점이라고 생각했기 때문입니다. 특히, 저의 경우 emotion에 대해 구체적으로 알지 못했고 사용해 본 경험이 없어, 선택을 할 때 emotion에 대한 의견을 내기 어려운 점이 있었습니다.

그러나 실제 배포 후 설치해 본 결과, Vite가 CSS 모듈로 생성된 고유 클래스명을 제대로 인식하지 못해 스타일이 적용되지 않는 문제를 발견했습니다. 다른 크루들도 같은 현상을 겪고 있었고, Vite가 해시된 클래스명을 의존성으로 인식하지 못하는 점이 원인으로 추측되고 있었습니다. 따라서 미션 마감 기한과 진행 상황을 고려했을 때, 이 문제를 빠르게 해결하기 위해 CSS 모듈 대신 emotion으로 전환하는 편이 효율적이라고 판단했고, 결과적으로 emotion을 선택하게 되었습니다.

저의 경우, 지식과 경험 부족으로 emotion을 사용할 근거는 명확하게 제시하기 어려웠습니다. 또한 CSS 모듈이 익숙하고 안정적이라고 생각했습니다. 다만, 이번 미션에서 처음 써 본 결과, emotion은 다음과 같은 장점이 있었습니다.

1. 조건부 스타일을 JS문법으로 간단하게 적용할 수 있다.
2. CSS를 위해 별도의 클래스 이름을 생성할 필요가 없어 템플릿이 깔끔해지고, 클래스 네이밍 충돌 등으로 인해 스타일이 잘못 적용되는 등의 실수를 할 가능성이 낮다.

이번 경험을 바탕으로, CSS 구현 방식을 선택할 때 emotion을 도입할 이유를 생각해 볼 수 있었던 것 같습니다.

JeLee-river commented on components/src/modal/Modal.tsx:
```diff
@@ -0,0 +1,137 @@
+  border: none;
+  border-radius: 8px;
+
+  margin-bottom: ${(props) => (props.isBottom ? 0 : null)};
+  width: ${(props) => (props.isBottom ? '100%' : null)};
+  max-width: ${(props) => (props.isBottom ? '100%' : null)};
+  border-bottom-left-radius: ${(props) => (props.isBottom ? 0 : null)};
+  border-bottom-right-radius: ${(props) => (props.isBottom ? 0 : null)};
```
오 좋은 방법인 것 같습니다. isBottom이 아닐때의 스타일을 분리하여 한 번에 분기처리를 하도록 수정해보겠습니다!

JeLee-river commented on hooks/src/lib/cardNumber/useCardNumbers.ts:
```diff
@@ -0,0 +1,81 @@
+  });
+
+  const checkIsNumber = (value: string) => {
+    const regex = /^[0-9]*$/;
```
문자열을 숫자로 변환한 값을 검증할 때 발생할 수 있는 문제를 최소화하고자 정규식을 사용했습니다!

`Number.isInteger`와 같이 숫자를 판별하는 메서드를 사용하기 위해서는 입력값을 숫자로 변환해야 합니다. 그러나 자바스크립트에서는 null이나 빈 문자열('')을 숫자 0으로 취급하기 때문에, 실제로 숫자가 아닌 값이 유효성 검사를 통과하는 가능성을 우려했습니다.
따라서 이러한 예외 상황을 방지하고자 변환 과정 없이 정규식으로 문자열을 바로 체크할 수 있는 방식을 선택했습니다.

JeLee-river committed:
style: bottom 모달과 default 모달의 스타일을 분리하여 한 번에 조건부 적용되도록 리팩토링

JeLee-river commented on hooks/src/lib/cardNumber/useCardNumbers.test.ts:
```diff
@@ -0,0 +1,34 @@
+
+describe('useCardNumbers', () => {
+  it('입력값이 정확히 업데이트 되어야 한다.', () => {
+    const { result } = renderHook(() => useCardNumbers());
```
beforeEach를 활용한다면 코드를 재사용할 수 있을 것 같습니다!

JeLee-river replied:
beforeEach를 활용하여 테스트마다 반복되는 로직을 재사용할 수 있도록 리팩토링했습니다.

[test: 반복되는 hook 렌더링 로직을 재사용하도록 개선 (d14eeae)](https://github.com/woowacourse/react-modules/pull/114/commits/d14eeaee6dc43b6103cb546651429f413d5ca0f0)

JeLee-river committed:
test: 반복되는 hook 렌더링 로직을 재사용하도록 개선

JeLee-river commented on hooks/src/lib/cardNumber/useCardNumbers.test.ts:
```diff
@@ -0,0 +1,34 @@
+import { CARD_NUMBER_ERROR_TYPES } from '../constants';
+
+describe('useCardNumbers', () => {
+  it('입력값이 정확히 업데이트 되어야 한다.', () => {
```
확실히 현재의 설명만으로는 어떤 동작을 테스트 하는 것인지 명확해보이지 않네요..! 좀 더 구체적으로 수정해보겠습니다!

JeLee-river committed:
test: 상태 변경 테스트의 설명을 구체적으로 수정

JeLee-river commented on hooks/src/lib/cardNumber/useCardNumbers.test.ts:
```diff
@@ -0,0 +1,34 @@
+    expect(result.current.cardNumbers.part1).toBe('1234');
+  });
+
+  it('입력값이 숫자가 아닐 때 isValid로 false를 반환하고 에러 메시지를 반환한다.', () => {
```
맞습니다! 테스트를 작성했을 당시에는 isValid와 메시지를 반환하는 구조였는데, 이후 리팩토링 과정에서 에러 타입을 반환하는 방식으로 변경되었습니다. 이를 테스트 설명에는 반영하지 못했네요. 😂  테스트 설명을 업데이트 하도록 하겠습니다. 코멘트 감사합니다!

JeLee-river replied:
테스트 동작과 일치하지 않는 설명을 수정했습니다. 이후, 유효성 검사 로직이 변경되면서 설명도 추가적으로 수정되었습니다!

[test: 실제 테스트 동작과 일치하지 않는 설명을 수정 (a4a1f96)](https://github.com/woowacourse/react-modules/pull/114/commits/a4a1f962109fa4cfb1dc25e1903d3d79b173e760)

[test: hook 수정에 따른 테스트 코드 수정 (ceedb81)](https://github.com/woowacourse/react-modules/pull/114/commits/ceedb8157ea3be03c53d2206c7b262a86ddfb3c6)

JeLee-river committed:
test: 실제 테스트 동작과 일치하지 않는 설명을 수정

JeLee-river committed:
refactor: 입력값의 유효성 검사 결과로 유효 여부(isValid)와 에러 타입을 반환하도록 수정

- 검사 결과가 유효한 경우 null을 반환했던 기존의 코드에서, isValid와 errorType을 반환하도록 수정하여 결과값의 의도가 명확히 드러나도록 함

JeLee-river committed:
test: hook 수정에 따른 테스트 코드 수정

JeLee-river commented on hooks/src/lib/cardNumber/useCardNumbers.ts:
```diff
@@ -0,0 +1,81 @@
+    return value.length <= 4;
+  };
+
+  const validateCardNumbers = (value: string) => {
+    const isNumber = checkIsNumber(value);
+    const isValidLength = checkIsValidLength(value);
+
+    if (!isNumber) {
+      return CARD_NUMBER_ERROR_TYPES.notNumber;
+    }
+
+    if (!isValidLength) {
+      return CARD_NUMBER_ERROR_TYPES.invalidLength;
+    }
+
+    return null;
+  };
```
유효성 검사 결과값의 의도가 좀 더 명확히 드러나도록 isValid 값과 에러 타입을 반환하도록 수정해보겠습니다!

JeLee-river replied:
유효성 검사 함수의 반환값을, 입력값의 유효성(isValid), 에러 타입을 반환하는 객체로 변경했습니다!

[변경 전]
```
if (!isNumber) {
      return CARD_NUMBER_ERROR_TYPES.notNumber;
}

return null
```

<br>

[변경 후]
```
if (!isNumber) {
      return { isValid: false, errorType: CARD_NUMBER_ERROR_TYPES.notNumber };
}

return { isValid: true };
```

[refactor: 입력값의 유효성 검사 결과로 유효 여부(isValid)와 에러 타입을 반환하도록 수정 (f6bdf24)](https://github.com/woowacourse/react-modules/pull/114/commits/f6bdf2403b60b2ac66efa8b9e0034a2b6986bafe)

JeLee-river commented on components/src/modal/Modal.tsx:
```diff
@@ -0,0 +1,137 @@
+    }
+  }, [open]);
+
+  useEffect(() => {
+    const handleClickOutside = (event: MouseEvent) => {
+      if (modalRef.current && event.target === modalRef.current) {
+        onClose();
+      }
+    };
+
+    if (open) {
+      document.addEventListener('mousedown', handleClickOutside);
+    }
+
+    return () => {
+      document.removeEventListener('mousedown', handleClickOutside);
+    };
+  }, [open, onClose]);
```
조사해본 결과, 현재의 코드에서 백드롭이 모달 컨테이너(최상위 요소)에 포함되는 점을 이용한다면 구현할 수 있습니다! 모달 컨테이너에 클릭 이벤트를 추가해 이벤트가 발생한 요소가 하위 요소가 아닌, 모달 컨테이너 요소 그 자체인지 판별하고, 모달을 닫도록 변경할 수 있을 것 같습니다!

추가로 유조가 ref와 document.addEventListener를 사용하지 않고 동작을 구현할 방법을 고민해보도록 하신 이유에 대해 생각해보았습니다. 클릭 이벤트와 같은 사용자와의 상호작용은 useEffect와 addEventListener를 이용하는 방식보다는 React에서 제공하는 이벤트 핸들러로 처리하는 것이 직관적이고 리액트스러운 코드이기 때문이 아닐까 추측했는데, 제가 생각한 내용이 맞을지 궁금합니다..!

JeLee-river committed:
refactor: 모달 외부를 클릭했을 때 모달이 닫히는 기능의 구현 방식 리팩토링

- useRef, addEventListener를 사용하던 방식에서 onClick 핸들러 함수를 이용하는 방식으로 수정

JeLee-river commented:
안녕하세요 유조! 연휴 잘 보내고 계시나요~? 😄 남겨주신 코멘트를 토대로 피드백 반영을 완료하여 재요청드립니다!

---

<br>

## 주요 변경 사항

주요 변경 사항은 다음과 같습니다.

<br>

### 1. 유효성 검사 결과의 반환 형식 변경

- 입력값의 유효성을 boolean으로 나타내는 `isValid`와, 에러 타입을 객체 형태로 반환하도록 수정함 [(f6bdf24)](https://github.com/woowacourse/react-modules/pull/114/commits/f6bdf2403b60b2ac66efa8b9e0034a2b6986bafe)

- 기존에 유효성 검사를 통과하면 null을 반환하던 방식은 결과를 직관적으로 이해하기 어렵기 때문에 이와 같이 변경함

<br>

### 2. 모달 외부를 클릭시 닫히는 기능의 구현 방식 변경

- useEffect 훅 안에서 document.addEventListener로 클릭 이벤트를 걸던 방식을, 모달 컨테이너에 onClick 핸들러를 직접 붙여 선언적으로 이벤트를 처리하도록 변경함 [(9a3e7c3)](https://github.com/woowacourse/react-modules/pull/114/commits/9a3e7c3a01a66d85ba838a629684a3fd83148fb2)

<br>

### 3. 모달의 default 스타일과 bottom 스타일을 분리

- 하나의 CSS 코드에서 분기 처리되고 있던 모달의 default/bottom 스타일을 분리하여 가독성 개선 [(3498ec5)](https://github.com/woowacourse/react-modules/pull/114/commits/3498ec51ed9352fc483dde6e484bb4dd23a5a15d)

---

감사합니다!

yujo11 commented on components/src/modal/Modal.tsx:
```diff
@@ -27,24 +28,16 @@ function ModalContainer({
-    if (open) {
-      document.addEventListener('mousedown', handleClickOutside);
+  const handleClickOutside = (event: React.MouseEvent<HTMLDialogElement>) => {
+    const { target } = event;
+    if (target === event.currentTarget) {
```
1. `event.target`과 `event.currentTarget`는 어떤 차이가 있을까요?
2. `target`은 `event`에서 꺼내서 사용하고, `currentTarget`은 꺼내서 사용하지 않는 이유가 있을까요?

JeLee-river replied:
1. event.target은 이벤트가 실제로 발생한 요소를 반환하지만, event.currentTarget은 이벤트가 등록된 요소를 반환합니다.
자바스크립트에서는 하위 요소에서 발생한 이벤트가 상위 요소까지 전파되기 때문에, 실제로 이벤트가 발생한 요소와 이벤트를 감지하도록 핸들러를 부착한 요소가 다를 수 있습니다. 

2. 엇 line 32에서 event를 구조 분해하고 있는 점을 인지하지 못했습니다. 😂 currentTarget도 같은 방식으로 이벤트 객체에서 추출할 수 있겠네요..! 2단계에서 반영하겠습니다!

yujo11 reviewed (approved):
안녕하세요 제나 ~~

코멘트 꼼꼼하게 확인하고 답변과 이모지 남겨주신 덕분에 제가 남겼던 코멘트가 잘 전달되고 있다는걸 확인할 수 있어서 좋았네요 ㅎㅎ

기존 코멘트들에 추가로 코멘트를 남긴 부분과 소소하게 궁금한 점에 대해서도 코멘트 남겼으니 같이 확인해주시면 좋겠네요!

이번 step은 여기서 이만 마치고 다음 step으로 넘어가면 좋을거 같습니다 ! 고생 많으셨습니다 제나 다음 step에서 다시 만나요 ~~  (코멘트를 함꼐 확인해준거 같은 재오도 같이 만나요 ㅎㅎ )

wmakerjun commented:
@coderabbitai
You are a senior React engineer.

# Task
From this pull request’s code, identify **exactly one recurring code smell or anti-pattern** whose removal will most improve long-term maintainability and correctness.

# How to respond
- **Pattern name**  
- **Where it appears** (files + line ranges or clear examples)  
- **Why it hurts** (short reason + link to React/JS best-practice doc if relevant)  
- **How to fix** (concise code snippet or pseudo-diff)  
- **Why this is the highest-leverage fix** (≤ 2 sentences)

# Style
Respond in bullet points, no extra tips or unrelated suggestions.

# Language
Please write your response **in Korean**.

JeLee-river commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one place** where the code is written in an imperative style, and explain how to make it **more declarative**.

# How to respond  
Start your response with the following markdown title:  
`# 1. ✨ 선언형 UI로서 개선할 피드백 포인트`

Then respond with:  
- **Imperative pattern name or symptom**
- **Where it appears** (files + line ranges or clear examples)  
- **Why it’s not declarative**  
- **AS-IS**: _Show the current (imperative) code snippet_  
- **TO-BE**: _Show the rewritten (declarative) version_  
- **Why this change improves React code quality** (≤ 2 sentences)

# Style  
Respond in bullet points, and include only one clear improvement point.

# Language  
Please write your response **in Korean**.

JeLee-river commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one state value** that can be **derived from other state or props**, and explain how removing it would simplify the code.

# How to respond  
Start your response with the following markdown title:  
`# 2. 📉 상태 최소화를 위한 개선 포인트`

Then respond with:  
- **Derived state name**  
- **Where it appears** (files + line ranges or variable name)  
- **Why it’s redundant**  
- **AS-IS**: _Show the redundant state and its logic_  
- **TO-BE**: _Show how to derive it from existing values_  
- **Why this is the highest-leverage fix** (≤ 2 sentences)

# Style  
Respond in bullet points, and focus on one state-related improvement.

# Language  
Please write your response **in Korean**.

JeLee-river commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one component** whose responsibility is unclear or mixed, and explain how to refactor it to **separate concerns more clearly**.

# How to respond  
Start your response with the following markdown title:  
`# 3. 🧩 컴포넌트 책임 분리를 위한 리팩토링 포인트`

Then respond with:  
- **Component name or file**  
- **Where the responsibility is unclear**  
- **Why this hurts reusability and maintainability**  
- **AS-IS**: _Show the current (overloaded) component logic_  
- **TO-BE**: _Show a refactored structure or extraction example_  
- **Why this is the most impactful structural change** (≤ 2 sentences)

# Style  
Respond in bullet points, and provide clear before/after contrast.

# Language  
Please write your response **in Korean**.

JeLee-river commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one recurring mistake, anti-pattern, or code smell** that appears in multiple places and seems to reflect the author's coding habit.  
Explain what it is, why it may keep happening, and how the author can recognize and change this pattern.

# How to respond  
Start your response with the following markdown title:  
`# 4. 🔁 반복적으로 나타나는 안티패턴 또는 코드 습관 피드백`

Then respond with:
- **Pattern or smell name**  
- **Where it appears repeatedly** (files + line ranges or variable/component patterns)  
- **Why it likely keeps happening** (assumed thought process or misconception)  
- **AS-IS**: _Example of current usage or structure_  
- **TO-BE**: _Refactored or improved structure with reasoning_  
- **How the author can self-check next time** (one reflection or checklist question)

# Style  
Respond in bullet points, focused on identifying **a single recurring behavioral pattern**.

# Language  
Please write your response **in Korean**.

JeLee-river commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one component** whose responsibility is unclear or mixed, and explain how to refactor it to **separate concerns more clearly**.

# How to respond  
Start your response with the following markdown title:  
`# 3. 🧩 컴포넌트 책임 분리를 위한 리팩토링 포인트`

Then respond with:  
- **Component name or file**  
- **Where the responsibility is unclear**  
- **Why this hurts reusability and maintainability**  
- **AS-IS**: _Show the current (overloaded) component logic_  
- **TO-BE**: _Show a refactored structure or extraction example_  
- **Why this is the most impactful structural change** (≤ 2 sentences)

# Style  
Respond in bullet points, and provide clear before/after contrast.

# Language  
Please write your response **in Korean**.

JeLee-river commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, evaluate how well it aligns with the principle that "React components should avoid unnecessary re-renders".  
If the code already prevents redundant re-renders effectively (e.g., through memoization, proper dependency arrays, or avoiding inline object/array creations), affirm that with a short explanation.  
If it could be improved, suggest one possible direction—but only if the improvement adds clear value without overengineering.

# How to respond  
Start your response with the following markdown title:  
# ✨ 한글 제목: 리렌더링 최적화 관점에서의 평가

Then respond with:
- ✅ **현재 코드가 원칙에 잘 맞는 이유** (if applicable)  
- ⚠️ **개선 여지가 있다면, 가능한 방향** (optional)  
- **AS-IS**: _Show the current code_
- **TO-BE**: _Show a refactored example_
- **스스로 점검할 수 있는 질문 하나**

# Style  
Use bullet points only. Do not force a suggestion—only provide one if it adds meaningful value.

# Language  
Please write your response **in Korean**.
